### PR TITLE
Log functional_collectives apis to distributed logger

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103288
* #103351

This logs functional collectives API calls with debug log level only.

(the `+` in the TORCH_LOGS cmdline enables debug level, otherwise only info level)

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Thanks for enabling TORCH_LOG!!

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -26,6 +27,8 @@ def is_torchdynamo_compiling():
         def is_torchdynamo_compiling():
             return False
 
+distributed_log = logging.getLogger(__name__)
```
Comment (ezyang) [inline torch/distributed/_functional_collectives.py:30]: Just call it `log`

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -376,6 +379,7 @@ def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str =
     :: N.B. If you pass a PG or a 1D list to perform a MPMD collective, the compiler won't be able to recover
     that information and perform collective algebraic optimization. Use other forms of input for that.
     """
+    distributed_log.debug("all_reduce size: %s, op: %s, group: %s, tag: %s", str(self.size()), reduceOp, group, tag)
```
Comment (ezyang) [inline torch/distributed/_functional_collectives.py:382]: don't call str in the argument list

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -464,6 +483,13 @@ def all_reduce_coalesced(self: List[torch.Tensor], reduceOp: str, group: RANK_TY
     :: N.B. If you pass a PG or a 1D list to perform a MPMD collective, the compiler won't be able to recover
     that information and perform collective algebraic optimization. Use other forms of input for that.
     """
+    distributed_log.debug(
+        "all_reduce_coalesced sizes: [%s], reduce_op: %s, group: %s, tag: %s",
+        ", ".join([str(t.size()) for t in self]),
```
Comment (ezyang) [inline torch/distributed/_functional_collectives.py:488]: don't do string formatting in the arg list. Use LazyString if you must.

Comment (wconstab) [reply torch/distributed/_functional_collectives.py]: hmm, is it still too bad to prepare a list of sizes without string formatting?

Comment (ezyang) [reply torch/distributed/_functional_collectives.py]: It's better. I personally draw the line at string formatting.

**Review by fduwjj — APPROVED** [review]
Nice!

**Comment by ezyang** [comment]
it's because ciflow/inductor didn't run on this job

**Comment by wconstab** [comment]
> it's because ciflow/inductor didn't run on this job

hmm, i thought that merge would run 'trunk' jobs including inductor, before finally merging? do i have to tag inductor manually if i want it to run before merge (when using `-i`)?

**Comment by ezyang** [comment]
no, sadly, inductor does not count as trunk

**Comment by wconstab** [comment]
hmm, the reason for the failure is an obvious problem I should have anticipated.  Dynamo is going to graph-break on these logs.  My only hope is if we can make dynamo not break when the log-level is set such that the log shouldn't print - not sure if that is easy to accomplish. cc @anijain2305 

alternatively, we could log deeper in the stack- inside the dispatcher op instead of inside the outer wrapper.  cc @wanchaol @kumpera

**Comment by wanchaol** [comment]
> 
> alternatively, we could log deeper in the stack- inside the dispatcher op instead of inside the outer wrapper. cc @wanchaol @kumpera

Yeah agreed, maybe we should just try logging in the dispatcher op instead (i.e. somewhere here https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L220), since dynamo would stop tracing into the aten op level

**Comment by wconstab** [comment]
What do you think about logging inside c10d layer vs logging in the traceable collectives layer? The thing that's annoying to consider is that depending if you are running eagerly or in compile, you'd get different log behavior at runtime unless the log is inside the c10d layer, since both eager and compiled traceable collectives both funnel into c10d.

**Comment by wanchaol** [comment]
> What do you think about logging inside c10d layer vs logging in the traceable collectives layer? The thing that's annoying to consider is that depending if you are running eagerly or in compile, you'd get different log behavior at runtime unless the log is inside the c10d layer, since both eager and compiled traceable collectives both funnel into c10d.

hmmm I see, I support adding to c10d as we can share the same logging with eager/compile

**Pull request closed without merging**